## Assignment 5
Nupoor Karnik

2. For parts (a) through (c), indicate which of i. through iv. is correct. Justify your answer.
(a) The lasso, relative to least squares, is:
i. More flexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance.
ii. More flexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.
iii. Less flexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance.
iv. Less flexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.

Ans: 
Option iii is correct: "Less flexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance."The lasso introduces a penalty term to the least squares objective function, which results in shrinkage of coefficient estimates towards zero, potentially leading to a simpler model with fewer predictors. This tends to increase bias but decrease variance.

(b) Repeat (a) for ridge regression relative to least squares.
Ans:
Option iii is also correct for ridge regression: "Less flexible and hence will give improved prediction accuracy when its increase in bias is less than its decrease in variance." The ridge regression introduces a penalty term, but it uses the L2 norm penalty instead of the L1 norm penalty used by the lasso. This also leads to shrinkage of coefficient estimates towards zero but usually does not result in exact zero coefficients. Ridge regression tends to increase bias and decrease variance.

(c) Repeat (a) for non-linear methods relative to least squares.
Ans:
Option ii is correct for non-linear methods: "More flexible and hence will give improved prediction accuracy when its increase in variance is less than its decrease in bias.". Non-linear methods introduce complexity by allowing for non-linear relationships between predictors and the response variable. These methods can capture more intricate patterns in the data but may also be prone to overfitting, leading to higher variance.

In [3]:
import pandas as pd
import numpy as np

#vix
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import OneHotEncoder

#load StatsModel API
import statsmodels.api as sm
import statsmodels.formula.api as smf

9. In this exercise, we will predict the number of applications received using the other variables in the College data set.
(a) Split the data set into a training set and a test set.

In [4]:
college = pd.read_csv('College.csv')

college.head()

Unnamed: 0 Private  Apps  Accept  Enroll  Top10perc  \
0  Abilene Christian University     Yes  1660    1232     721         23   
1            Adelphi University     Yes  2186    1924     512         16   
2                Adrian College     Yes  1428    1097     336         22   
3           Agnes Scott College     Yes   417     349     137         60   
4     Alaska Pacific University     Yes   193     146      55         16   

   Top25perc  F.Undergrad  P.Undergrad  Outstate  Room.Board  Books  Personal  \
0         52         2885          537      7440        3300    450      2200   
1         29         2683         1227     12280        6450    750      1500   
2         50         1036           99     11250        3750    400      1165   
3         89          510           63     12960        5450    450       875   
4         44          249          869      7560        4120    800      1500   

   PhD  Terminal  S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0   70        78       18.1           12    7041         60  
1   29        30       12.2           16   10527         56  
2   53        66       12.9           30    8735         54  
3   92        97        7.7           37   19016         59  
4   76        72       11.9            2   10922         15

In [5]:
college = college.rename(columns={'Unnamed: 0': 'Name'})
college['Private'] = college['Private'].map({'Yes': 1, 'No': 0})

In [6]:
college.head()

Name  Private  Apps  Accept  Enroll  Top10perc  \
0  Abilene Christian University        1  1660    1232     721         23   
1            Adelphi University        1  2186    1924     512         16   
2                Adrian College        1  1428    1097     336         22   
3           Agnes Scott College        1   417     349     137         60   
4     Alaska Pacific University        1   193     146      55         16   

   Top25perc  F.Undergrad  P.Undergrad  Outstate  Room.Board  Books  Personal  \
0         52         2885          537      7440        3300    450      2200   
1         29         2683         1227     12280        6450    750      1500   
2         50         1036           99     11250        3750    400      1165   
3         89          510           63     12960        5450    450       875   
4         44          249          869      7560        4120    800      1500   

   PhD  Terminal  S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0   70        78       18.1           12    7041         60  
1   29        30       12.2           16   10527         56  
2   53        66       12.9           30    8735         54  
3   92        97        7.7           37   19016         59  
4   76        72       11.9            2   10922         15

In [20]:
# Assuming 'applications' is the target variable and other variables are predictors
X = college.drop(['Apps', 'Name'], axis=1)  # Features
y = college['Apps']  # Target variable

# Split the dataset into train and test sets (80% train, 20% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Optional: If you want to see the shapes of train and test sets
print("Training set:", X_train.shape, y_train.shape)
print("Test set:", X_test.shape, y_test.shape)

Training set: (621, 17) (621,)
Test set: (156, 17) (156,)


(b) Fit a linear model using least squares on the training set, and report the test error obtained.

In [26]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# Fit a linear model using least squares on the training set
model = LinearRegression(fit_intercept=True)
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Calculate the test error (Mean Squared Error)
test_error = mean_squared_error(y_test, y_pred)

# Report the test error
print("Test Error(Linear): " + str(test_error))

Test Error(Linear): 1492443.3790390575


(c) Fit a ridge regression model on the training set, with λ chosen by cross-validation. Report the test error obtained.

In [27]:
from sklearn.linear_model import RidgeCV


# Fit a ridge regression model with cross-validation
ridge_model = RidgeCV(alphas=[0.1, 1.0, 10.0], cv=5)  # You can adjust the alphas as needed
ridge_model.fit(X_train, y_train)

# Predict on the test set
y_pred = ridge_model.predict(X_test)

# Calculate the test error (Mean Squared Error)
test_error = mean_squared_error(y_test, y_pred)

# Report the test error
print("Test Error (Ridge Regression):", test_error)

Test Error (Ridge Regression): 1478572.8112796955


(d) Fit a lasso model on the training set, with λ chosen by cross- validation. Report the test error obtained, along with the number of non-zero coefficient estimates.

In [28]:
from sklearn.linear_model import LassoCV

# Fit a Lasso model with cross-validation
lasso_model = LassoCV(cv=5)  
lasso_model.fit(X_train, y_train)

# Predict on the test set
y_pred = lasso_model.predict(X_test)

# Calculate the test error (Mean Squared Error)
test_error = mean_squared_error(y_test, y_pred)

# Count the number of non-zero coefficient estimates
num_non_zero_coefs = np.sum(lasso_model.coef_ != 0)

# Report the test error and number of non-zero coefficient estimates
print("Test Error (Lasso Regression):", test_error)
print("Number of Non-Zero Coefficients:", num_non_zero_coefs)

Test Error (Lasso Regression): 1587020.0176529188
Number of Non-Zero Coefficients: 7


(g) Comment on the results obtained. How accurately can we predict the number of college applications received? Is there much difference among the test errors resulting from these five approaches?

From the test errors obtained:

- Linear Regression: 1492443.3790390575
- Ridge Regression: 1478572.8112796955
- Lasso Regression: 1587020.0176529188

It can be seen that the Ridge Regression model performs slightly better than the Linear Regression model in terms of test error, while the Lasso Regression model has the highest test error among the three. Additionally, the Lasso Regression model has only 7 non-zero coefficient estimates, indicating that it has performed some feature selection by effectively shrinking coefficients to zero.

Regarding the accuracy of prediction, the test errors obtained are relatively high, indicating that there is considerable variability in predicting the number of college applications received using the given features. The difference among the test errors resulting from these approaches is not substantial, but the Ridge Regression model seems to have a slight advantage over Linear Regression, possibly due to its ability to handle multicollinearity and overfitting by introducing a penalty term.

Overall, while the models can provide some predictive value, the test errors suggest that accurately predicting the number of college applications received based on the given features may be challenging, and there might be other factors not considered in the analysis that influence application numbers. Further exploration and feature engineering may be needed to improve predictive accuracy.